File: leetcode/easy/704_binary_search/binary_search.py
from typing import List


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        return self.rec(nums, target, 0, len(nums))

    def rec(self, nums, target, left, right):
        mid = left + (right - left) // 2
        if nums[mid] == target:
            return mid
        if right - left <= 1:
            return -1
        if nums[mid] < target:
            return self.rec(nums, target, mid, right)
        if nums[mid] > target:
            return self.rec(nums, target, left, mid)


Solution().search([-1, 0, 3, 5, 9, 12], 9)

File: leetcode/easy/704_binary_search/binary_search.cpp
#include <vector>

class Solution {
public:
  int search(std::vector<int> &nums, int target) {
    int left = 0, right = nums.size() - 1;
    while (true) {
      if (nums[left] == target) {
        return left;
      }
      if (nums[right] == target) {
        return right;
      }
      if (left >= right - 1) {
        return -1;
      }
      const int mid = left + (right - left) / 2;
      if (target > nums[mid]) {
        left = mid;
      } else {
        right = mid;
      }
    }
  }
};
